File: psets/pset1/credit.c
// https://cs50.harvard.edu/x/2022/psets/1/credit/

#include <stdio.h>
#include <cs50.h>
#include <math.h>

// Returns the nth digit of a number from right to left
int get_nth_digit(long number, int n);

int main(void)
{
    long int credit_card = get_long("Number: ");

    // Calculate the number of digits
    int digits = log10(credit_card) + 1;

    // Implementation of Luhn’s algorithm
    int odd_index_sum = 0, even_index_sum = 0;
    for (int i = 0; i < digits; i++)
    {
        int current_digit = get_nth_digit(credit_card, i + 1);

        if (i % 2 == 0)
        {
            even_index_sum += current_digit;
        }
        else
        {
            int digit_doubled = current_digit * 2;
            odd_index_sum += (digit_doubled / 10) + (digit_doubled % 10);
        }
    }
    bool valid_card = (even_index_sum + odd_index_sum) % 10 == 0;

    // Assigns the company that owns the credit card to 'company'
    string company = "INVALID";
    if (valid_card)
    {
        int first_digit = get_nth_digit(credit_card, digits);
        int second_digit = get_nth_digit(credit_card, digits - 1);

        if (digits == 13 && first_digit == 4)
        {
            company = "VISA";
        }
        else if (digits == 15 && first_digit == 3 && (second_digit == 4 || second_digit == 7))
        {
            company = "AMEX";
        }
        else if (digits == 16)
        {
            if (first_digit == 4)
            {
                company = "VISA";
            }
            else if (first_digit == 5 && second_digit >= 1 && second_digit <= 5)
            {
                company = "MASTERCARD";
            }
        }
    }

    // Prints 'company'
    printf("%s\n", company);
}

int get_nth_digit(long number, int n)
{
    return (number % (long)pow(10, n)) / pow(10, n - 1);
}

File: psets/pset6/credit.py
# https://cs50.harvard.edu/x/2022/psets/6/credit/

# Importing libraries
from re import match
from cs50 import get_string


def main():
    number = get_string("Number: ")

    # Filter user input
    while not number.isdigit():
        number = get_string("Number: ")

    # Default value
    card = "INVALID"

    if match(r"^\d{13}$|^\d{15,16}$", number):
        if luhns_algorithm(number):
            length = len(number)

            # Match card company
            if length == 13:
                if number[0] == '4':
                    card = "VISA"
            elif length == 15:
                if number[0] == '3':
                    if number[1] in ['4', '7']:
                        card = "AMEX"
            elif length == 16:
                if number[0] == '4':
                    card = "VISA"
                elif number[0] == '5':
                    if int(number[1]) in range(1, 6):
                        card = "MASTERCARD"

    print(card)


# Luhn's algorithm implementation
def luhns_algorithm(number):
    sumEven = sumOdd = 0

    for i, c in enumerate(reversed(number)):
        if i % 2 == 1:
            n = int(c) * 2
            sumOdd += n // 10 + n % 10
        else:
            sumEven += int(c)

    result = sumEven + sumOdd

    return result % 10 == 0


# Defining main
if __name__ == "__main__":
    main()
